### PR TITLE
DOC: Add S3DataGrabber example

### PR DESCRIPTION
**Pull request by satra**
this is to improve usage of s3 data grabber.
several examples here: https://github.com/kellyhennigan/cueexp_scripts/blob/25372d4f221762af6d79d1af67a8570f2c546b20/svmrfe/venv/lib64/python2.7/site-packages/nipype/interfaces/tests/test_io.py

## Acknowledgment

- [X] \(Mandatory\) I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

**Comment by satra** [comment]
@TimVanMourik ( here is an example of s3 datagrabber )

File: nipype/interfaces/io.py [hunk 1]
```diff
@@ -829,7 +829,22 @@ class S3DataGrabber(LibraryBaseInterface, IOBase):
         "template" uses regex style formatting, rather than the
         glob-style found in the original DataGrabber.
 
+        Examples
+        --------
+
+        >>> s3grab = S3DataGrabber(infields=['subj_id'], outfields=["func", "anat"])
+        >>> s3grab.inputs.bucket = 'openneuro'
+        >>> s3grab.inputs.sort_filelist = True
+        >>> s3grab.inputs.template = '*'
```
Comment (effigies) [inline nipype/interfaces/io.py]:
```suggestion
        >>> s3grab.inputs.template = '*'
```

File: nipype/interfaces/io.py [hunk 2]
```diff
@@ -829,7 +829,22 @@ class S3DataGrabber(LibraryBaseInterface, IOBase):
...
 
+        Examples
+        --------
+
+        >>> s3grab = S3DataGrabber(infields=['subj_id'], outfields=["func", "anat"])
+        >>> s3grab.inputs.bucket = 'openneuro'
+        >>> s3grab.inputs.sort_filelist = True
+        >>> s3grab.inputs.template = '*' 
+        >>> s3grab.inputs.anon = True
+        >>> s3grab.inputs.bucket_path = 'ds000101/ds000101_R2.0.0/uncompressed/'
```
Comment (effigies) [inline nipype/interfaces/io.py]:
```suggestion
        >>> s3grab.inputs.bucket_path = 'ds000101/ds000101_R2.0.0/uncompressed/'
```

File: nipype/interfaces/io.py [hunk 3]
```diff
@@ -829,7 +829,22 @@ class S3DataGrabber(LibraryBaseInterface, IOBase):
...
+        Examples
+        --------
+
+        >>> s3grab = S3DataGrabber(infields=['subj_id'], outfields=["func", "anat"])
+        >>> s3grab.inputs.bucket = 'openneuro'
+        >>> s3grab.inputs.sort_filelist = True
+        >>> s3grab.inputs.template = '*' 
+        >>> s3grab.inputs.anon = True
+        >>> s3grab.inputs.bucket_path = 'ds000101/ds000101_R2.0.0/uncompressed/' 
+        >>> s3grab.inputs.local_directory = '/tmp'
```
Comment (effigies) [inline nipype/interfaces/io.py]:
```suggestion
        >>> s3grab.inputs.local_directory = '/tmp'
```

File: nipype/interfaces/io.py [hunk 4]
```diff
@@ -829,7 +829,22 @@ class S3DataGrabber(LibraryBaseInterface, IOBase):
...
+        --------
+
+        >>> s3grab = S3DataGrabber(infields=['subj_id'], outfields=["func", "anat"])
+        >>> s3grab.inputs.bucket = 'openneuro'
+        >>> s3grab.inputs.sort_filelist = True
+        >>> s3grab.inputs.template = '*' 
+        >>> s3grab.inputs.anon = True
+        >>> s3grab.inputs.bucket_path = 'ds000101/ds000101_R2.0.0/uncompressed/' 
+        >>> s3grab.inputs.local_directory = '/tmp' 
+        >>> s3grab.inputs.field_template = {'anat': '%s/anat/%s_T1w.nii.gz', 'func': '%s/func/%s_task-simon_run-1_bold.nii.gz'}
```
Comment (effigies) [inline nipype/interfaces/io.py]:
```suggestion
        >>> s3grab.inputs.field_template = {'anat': '%s/anat/%s_T1w.nii.gz', 'func': '%s/func/%s_task-simon_run-1_bold.nii.gz'}
```

File: nipype/interfaces/io.py [hunk 5]
```diff
@@ -829,7 +829,22 @@ class S3DataGrabber(LibraryBaseInterface, IOBase):
...
+
+        >>> s3grab = S3DataGrabber(infields=['subj_id'], outfields=["func", "anat"])
+        >>> s3grab.inputs.bucket = 'openneuro'
+        >>> s3grab.inputs.sort_filelist = True
+        >>> s3grab.inputs.template = '*' 
+        >>> s3grab.inputs.anon = True
+        >>> s3grab.inputs.bucket_path = 'ds000101/ds000101_R2.0.0/uncompressed/' 
+        >>> s3grab.inputs.local_directory = '/tmp' 
+        >>> s3grab.inputs.field_template = {'anat': '%s/anat/%s_T1w.nii.gz', 'func': '%s/func/%s_task-simon_run-1_bold.nii.gz'} 
+        >>> s3grab.inputs.template_args = {'anat': [['subj_id', 'subj_id']], 'func': [['subj_id', 'subj_id']]}
```
Comment (effigies) [inline nipype/interfaces/io.py]:
```suggestion
        >>> s3grab.inputs.template_args = {'anat': [['subj_id', 'subj_id']], 'func': [['subj_id', 'subj_id']]}
```

**Comment by wtriplett** [comment]
Thanks for this example. The `s3grab.inputs.template = '*'` suggests a glob style wildcard, while the docstring claims (correctly) that for the S3DataGrabber a regex is used. So while this example works, maybe a more regex-looking template spec would highlight that difference from `DataGrabber` to the reader.

**Comment by effigies** [comment]
Going to go ahead and merge this. @wtriplett Feel free to open a pull request with an improved (or, better, another) example.

**Comment by satra** [comment]
@timvanmourik were you able to use this to create an s3datagrabber example? i'm not completely sure how i would initialize the s3datagrabber appropriately with giraffe.

@wtriplett - it would be nice to add a regex example. i just wanted to get an example in, since there were none :)

**Comment by TimVanMourik** [comment]
Sorry for the delay. Yes, made an example:
https://giraffe.tools/porcupine/TimVanMourik/SomeGiraffeExample/s3 or [a permanent link to current status](https://giraffe.tools/porcupine/TimVanMourik/SomeGiraffeExample/9a436be5e312f7a5b68acc41cfeb73fce3b2da44), in case the branch gets updated.

This example:
* It implements the file handling by:
  * mapping the input `subj_id` and `run_num` ...
  * ...via the `field_template` and `template_args`...
  * ...to output `func` and `struct`. 
  *  It's far from intuitive but it's the way the `S3DataGrabber` works so there's not much I can do about it from my side.
* Grabs data from the example file you mentioned
  * Iterates over subjects in the first node, but not over runs: the structural pipeline needs to be performed only for each subject. The functional pipeline is later also iterated over runs.
* Further implements a basic pipeline and stores the data at the end.

All Nipype modules that requires dynamic inputs/outputs are a bit cumbersome in GiraffeTools, but can be dealt with by adding ports manually. It is a bit contrary to the philosophy though, as I like to think of functions as static.

**Pull request merged**